processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java

```diff
@@ -0,0 +1,313 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby;
+
+import com.google.common.base.Strings;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import io.druid.data.input.Row;
+import io.druid.math.expr.Evals;
+import io.druid.math.expr.Expr;
+import io.druid.math.expr.Parser;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.FloatColumnSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.NumericColumnSelector;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.Column;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ColumnCapabilitiesImpl;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.data.IndexedInts;
+import it.unimi.dsi.fastutil.ints.IntIterator;
+import it.unimi.dsi.fastutil.ints.IntIterators;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class RowBasedColumnSelectorFactory implements ColumnSelectorFactory
+{
+  private final Supplier<? extends Row> row;
+  private final Map<String, ValueType> columnTypes;
+
+  public RowBasedColumnSelectorFactory(final Supplier<? extends Row> row, final Map<String, ValueType> columnTypes)
+  {
+    this.row = row;
+    this.columnTypes = columnTypes != null ? columnTypes : ImmutableMap.<String, ValueType>of();
+  }
+
+  public static RowBasedColumnSelectorFactory create(
+      final Supplier<? extends Row> row,
+      final Map<String, ValueType> columnTypes
+  )
+  {
+    return new RowBasedColumnSelectorFactory(row, columnTypes);
+  }
+
+  public static RowBasedColumnSelectorFactory create(
+      final ThreadLocal<? extends Row> row,
+      final Map<String, ValueType> columnTypes
+  )
+  {
+    return new RowBasedColumnSelectorFactory(
+        new Supplier<Row>()
+        {
+          @Override
+          public Row get()
+          {
+            return row.get();
+          }
+        },
+        columnTypes
+    );
+  }
+
+  // This dimension selector does not have an associated lookup dictionary, which means lookup can only be done
```

Comment:
The first sentence is accurate; the second sentence is kind-of accurate but also probably distracting. I deleted the second sentence and moved the first sentence to `getValueCardinality` where it would make more sense.